drv/mb85rsxx-fram/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+    pub fn write_disable(self) -> Result<(), (SpiError, Self)> {
+        match self.0.do_write_disable() {
+            Ok(_) => {
+                // Don't do it again.
+                core::mem::forget(self);
+                Ok(())
+            }
+            Err(e) => Err((e, self)),
+        }
+    }
+}
+
+impl FramId {
+    const MANUFACTURER_FUJITSU: u8 = 0x04;
+
+    fn read<S: SpiServer>(spi: &SpiDevice<S>) -> Result<Self, SpiError> {
+        // Indicates that we must read another two bytes to get the product ID.
+        const CONTINUE: u8 = 0x7f;
+
+        let _cs_is_held_low_while_this_exists = spi
+            .lock_auto(CsState::Asserted)
+            .map_err(|_| SpiError::TaskRestarted)?;
+        let mut buf = [0; 3];
+        spi.exchange(&[Opcode::ReadId as u8, 0, 0], &mut buf)?;
+        let [_, mfg_id, cont] = buf;
+
+        let product_id = if cont == CONTINUE {
+            let mut buf = [0; 2];
+            spi.read(&mut buf)?;
+            u16::from_be_bytes(buf)
+        } else {
+            // no continuation code --- use the bytes we just read.
+            // AFAICT, the Fujitsu FRAM chips don't do this, but other
+            // manufacturers' FRAM chips that use the same protocol do? Would
+            // have to read some more datasheets to be sure.
+            u16::from_be_bytes([mfg_id, cont])
+        };
+
+        Ok(Self { mfg_id, product_id })
+    }
+}
+
+impl<S: SpiServer, const ID: u16> Drop for WritableFram<'_, S, { ID }> {
+    fn drop(&mut self) {
+        // Put the FRAM back the way we found it.
+        let _ = self.0.do_write_disable();
+    }
+}
+
+impl From<SpiError> for FramInitError {
+    fn from(value: SpiError) -> Self {
+        FramInitError::SpiError(value)
+    }
+}
+
+/// Product IDs for various Fujitsu FRAM devices.
+///
+/// This driver primarily uses the product ID to determine whether the size of
+/// the FRAM chip matches the expected size. Other features of the FRAM chip can
+/// also be determined from the product ID, as well; however, for the purposes
+/// of this driver, we basically only care about the size of the device, because
+/// we don't use some of the fancier features, such as bank protection or the
+/// HOLD_L and WP_L pins.
+pub mod product_id {
+
+    /// 2kb Fujitsu FRAM
+    pub const MB85RS16: u16 = 0x0101;
+    /// 8kb Fujitsu FRAM
+    pub const MB85RS64V: u16 = 0x0302;
+    /// 8kb Fujitsu FRAM
+    pub const MB85RS64T: u16 = 0x2303;
+    /// 32kb Fujitsu FRAM
+    pub const MB85RS256TY: u16 = 0x2503;
+    /// 128kb Fujitsu FRAM
+    pub const MB85RS1MT: u16 = 0x2703;
+    /// 256kb Fujitsu FRAM
+    pub const MB85RS2MTA: u16 = 0x4803;
+    /// 512kb Fujitsu FRAM
+    pub const MB85RS4MT: u16 = 0x4903;
+
+    /// Returns the size in bytes of the FRAM chip, based on its product ID.
+    pub(super) const fn size(product_id: u16) -> usize {
+        // The first 5 bits of the product ID give the density of the FRAM chip
+        // in multiples of 2KiB.
```

Comment:
**question:** At the risk of sounding stupid: Is "multiples of" correct? I would read eg. 0x3 to become 3 * 2 KiB, not the 2 ^ 3 * 1 KiB that it actually means.

Reply:
yeah, i think i should have said..."exponents of two", or something, here? but "exponents of two KiB" also sounds wrong, it sounds like it's saying `2KiB ^ 3`, which is also not what it is...

Reply:
Maybe "gives us the kilobyte density of [] in powers of two" or "gives us the density of [] in kilobyte powers of two"?